client/routing.js

```diff
@@ -232,134 +229,167 @@ const isExecuted = receipt => {
-      log.debug(`[${txRequest.address}] Already called.`)
-      cleanup(conf, txRequest)
-      return 7
-    }
-    if ((await txRequest.inReservedWindow()) && txRequest.isClaimed && !isClaimedByUs(conf, txRequest)) {
-      return 8
-    }
-    // This hacks the cache to set all executed requests to store the value
-    // of -1 if it has been executed.
-    if (conf.cache.get(txRequest.address) <= 101) {
-      log.debug(`[${txRequest.address}] Already executed.`)
-      return 9
+  return next
+}
+
+const executionState = async(conf, txRequest) => {
+  const log = conf.logger
+  const { web3 } = conf
+  const self = STATE.EXECUTION
+  const next = STATE.DONE
+
+  if (txRequest.wasCalled) {
+    log.debug(`[${txRequest.address}] Already called.`)
+    return next
+  }
+
+  if ((await txRequest.inReservedWindow()) && !isClaimedByUs(conf, txRequest)) {
+    return self
+  }
+
+  try {
+    const { receipt, from } = await execute(conf, txRequest)
+
+    if (receipt && receipt.status == 1) {
+      if (isExecuted(receipt)) {
+        const data = receipt.logs[0].data
+        const timeBounty = web3.toDecimal(data.slice(0, 66))
+
+        log.info(`[${txRequest.address}] Executed.`)
+
+        conf.statsdb.updateExecuted(from, timeBounty, 0)
+
+        return next
+      } else {
+        log.info(`[${txRequest.address}] Execution failed. Transaction already executed.`)
+      }
+    } else {
+      log.error(`[${txRequest.address}] Execution failed.`)
     }
-    execute(conf, txRequest)
-      .then(result => {
-        const { web3 } = conf
-        const { receipt, from } = result
-
-        if (receipt && receipt.status == 1) {
-          if (isExecuted(receipt)) {
-              const data = receipt.logs[0].data
-              const timeBounty = web3.toDecimal(data.slice(0, 66))
-
-              log.info(`[${txRequest.address}] Executed.`)
-              conf.cache.set(txRequest.address, 100)
-              conf.statsdb.updateExecuted(from, timeBounty, 0)
-              
-              return
-            } else {
-              log.info(`[${txRequest.address}] Execution failed. Transaction already executed.`)
-            }
-        } else {
-          log.error(`[${txRequest.address}] Execution failed.`)
-        }
 
-        const txCost = receipt.gasUsed * txRequest.data.txData.gasPrice
-        conf.statsdb.updateExecuted(from, 0, txCost)
-      })
-      .catch(err => log.error(err))
-    return 10
+    const txCost = receipt.gasUsed * txRequest.data.txData.gasPrice
+    conf.statsdb.updateExecuted(from, 0, txCost)
+  } catch (err) {
+    log.error(err)
   }
 
-  // If the transaction request is expired, we try to clean it
-  if (await txRequest.afterExecutionWindow()) {
-    log.debug(`[${txRequest.address}] Cleaning up expired txRequest and removing from cache.`)
-    cleanup(conf, txRequest)
-    return 11
+  return next
+}
+
+const STATE = {
+  PRE_CLAIMING: 0,
+  CLAIMING: 1,
+  PRE_EXECUTION: 2,
+  EXECUTION: 3,
+  DONE: 4,
+}
```

Comment:
Maybe put the STATE constant at the top of the file? Since it's being referenced throughout the file.

Reply:
+1